pex/jobs.py

```diff
@@ -595,3 +614,195 @@ def spawned_jobs():
+        pool = cast("Pool", context.Pool(processes=size))
+    except (AttributeError, ValueError):
+        pool = multiprocessing.Pool(processes=size)
+
+    try:
+        yield pool
+    finally:
+        pool.close()
+        pool.join()
+
+
+# This is derived from experiment and backed up by multiprocessing internal default chunk size of
+# 4 for chunked mapping. The overhead of setting up and bookkeeping the multiprocessing pool
+# processes and communication pipes seems to be worth if at least 4 items are processed per slot.
+MULTIPROCESSING_DEFAULT_MIN_AVERAGE_LOAD = 4
+
+
+def imap_parallel(
+    inputs,  # type: Iterable[_I]
+    function,  # type: Callable[[_I], _O]
+    max_jobs=None,  # type: Optional[int]
+    min_average_load=MULTIPROCESSING_DEFAULT_MIN_AVERAGE_LOAD,  # type: int
+    costing_function=None,  # type: Optional[Callable[[_I], Comparable]]
+    result_render_function=None,  # type: Optional[Callable[[_O], Any]]
+    verb="process",  # type: str
+    verb_past="processed",  # type: str
+):
+    # type: (...) -> Iterator[_O]
+    """Enhanced version of `multiprocessing.Pool.imap` that optimizes pool size and input ordering.
+
+    :param inputs: The items to process with `function`.
+    :param function: A function that takes a single argument from `inputs` and returns a result.
+    :param max_jobs: The maximum number of Python processes to spawn to service the `inputs`.
+    :param min_average_load: The minimum avg. number of inputs each Python process should service.
+    :param costing_function: A function that can estimate the cost of processing each input.
+    :param result_render_function: A function that can take a result from `function` and render an
+                                   identifier for it.
+    :param verb: A verb indicating what the function does; "process" by default.
+    :param verb_past: The past tense of `verb`; "processed" by default.
+    :return: An iterator over the mapped results.
+    """
+    input_items = list(inputs)
+    if not input_items:
+        return
+
+    if costing_function is not None:
+        # We ensure no job slot is so unlucky as to get all the biggest jobs and thus become an
+        # un-necessarily long pole by sorting based on cost. Some examples to illustrate the effect
+        # using 6 input items and 2 job slots:
+        #
+        # 1.) Random worst case ordering:
+        #         [9, 1, 1, 1, 1, 10] -> slot1[9] slot2[1, 1, 1, 1, 10]: 14 long pole
+        #     Sorted becomes:
+        #         [10, 9, 1, 1, 1, 1] -> slot1[10, 1, 1] slot2[9, 1, 1]: 12 long pole
+        # 2.) Random worst case ordering:
+        #         [6, 4, 3, 10, 1, 1] -> slot1[6, 10] slot2[4, 3, 1, 1]: 16 long pole
+        #     Sorted becomes:
+        #         [10, 6, 4, 3, 1, 1] -> slot1[10, 3] slot2[6, 4, 1, 1]: 13 long pole
+        #
+        input_items.sort(key=costing_function, reverse=True)
```

Comment:
There is now new binning logging below and it just proves this out. For example, for the `pantsbuild.pants==2.17.1` example case:
```
# Unsorted
pex: Elapsed time per install job:
    [657197] 3.75s 5 wheels
    [657199] 4.33s 7 wheels
    [657198] 4.43s 2 wheels
    [657200] 7.39s 3 wheels
    [657196] 10.13s 5 wheels

# Sorted biggest 1st
pex: Elapsed time per install job:
    [656446] 4.36s 8 wheels
    [656447] 4.36s 9 wheels
    [656448] 4.49s 3 wheels
    [656444] 7.04s 1 wheel
    [656445] 7.86s 1 wheel
```
This ~2.5s improvement in the overall processing time was consistent.

Reply:
To be clear - this was not new behavior, just new experimental confirmation of what was just previously chicken-scratch with pen and paper proving this to myself in the initial PR.